File: sdk/node/src/api/course.content/course.content.ts

import { CourseContentCreateModel, CourseContentUpdateModel, CourseContentSearchParams, ApiResponse } from './course.content.model';
import { APIClient } from '../client';
import { plainToInstance } from 'class-transformer';
import { validateSync } from 'class-validator';
import { faker } from '@faker-js/faker';

export class CourseContent {
    constructor(private apiClient: APIClient) {}

    private validateDataclass<T extends object>(data: Record<string, any>, DataClass: new () => T): T 
    {
        const instance = plainToInstance(DataClass, data as Partial<T>); 
        const errors = validateSync(instance);
        if (errors.length > 0) {
            throw new Error(`Validation failed: ${JSON.stringify(errors)}`);
        }
        return instance;
    }

    public async create(data: Record<string, any>): Promise<ApiResponse<CourseContentCreateModel>> 
    {
        const validatedData = this.validateDataclass(data, CourseContentCreateModel);
        return this.apiClient.makeRequest<ApiResponse<CourseContentCreateModel>>('POST', 'course_contents', validatedData);
    }

    public async getById(course_contents_Id: string): Promise<ApiResponse<CourseContentCreateModel>> 
    {
        if (typeof course_contents_Id !== 'string') {
            throw new TypeError(`Expected 'course_contents_Id' to be a string, got ${typeof course_contents_Id}.`);
        }
        return this.apiClient.makeRequest<ApiResponse<CourseContentCreateModel>>('GET', `course_contents/${ course_contents_Id }`);
    }

    public async search(queryParams: Record<string, any>): Promise<ApiResponse<CourseContentCreateModel[]>> 
    {
        const validatedData = this.validateDataclass(queryParams, CourseContentSearchParams);
        return this.apiClient.makeRequest<ApiResponse<CourseContentCreateModel[]>>('GET', `course_contents/search`, validatedData);
    }

    public async update(course_contents_Id: string, data: Record<string, any>): Promise<ApiResponse<CourseContentCreateModel>> 
    {
        if (typeof course_contents_Id !== 'string') {
            throw new TypeError(`Expected 'course_contents_Id' to be a string, got ${typeof course_contents_Id}.`);
        }
        const validatedData = this.validateDataclass(data, CourseContentUpdateModel);
        return this.apiClient.makeRequest<ApiResponse<CourseContentCreateModel>>('PUT', `course_contents/${ course_contents_Id }`, validatedData);
    }

    public async delete(course_contents_Id: string): Promise<ApiResponse<void>> 
    {
        if (typeof course_contents_Id !== 'string') {
            throw new TypeError(`Expected 'course_contents_Id' to be a string, got ${typeof course_contents_Id}.`);
        }
        return this.apiClient.makeRequest<ApiResponse<void>>('DELETE', `course_contents/${ course_contents_Id }`);
    }

    public async courseContentOperations(): Promise<void> 
    {
        try {
            const courseContentCreateData = {
            "Title": faker.lorem.sentence(3),
            "Description": faker.lorem.words(3),
            "ImageUrl": faker.internet.url(),
            "DurationInMins": faker.number.int({ max: 9999 }),
            "ContentType": "Text",
            "ResourceLink": faker.lorem.sentence(3),
            "ActionTemplateId": "{{ACTION_TEMPLATE_ID}}",
            "Sequence": faker.number.int({ max: 9999 }),

            };

            const courseContentCreate = await this.create(courseContentCreateData);
            console.log('Create:', JSON.stringify(courseContentCreate.Data, null, 2));

            const course_contents_Id = courseContentCreate.Data?.id; 
            if (!course_contents_Id) throw new Error('CourseContent ID not returned.');

            const courseContentGetById = await this.getById(course_contents_Id);
            console.log('GetById:', JSON.stringify(courseContentGetById.Data, null, 2));

            const courseContentSearchData = {
            "Title": faker.lorem.sentence(3),
            "Description": faker.lorem.words(3),
            "ImageUrl": faker.internet.url(),
            "DurationInMins": faker.number.int({ max: 9999 }),
            "ContentType": "Text",
            "ResourceLink": faker.lorem.sentence(3),
            "ActionTemplateId": "{{ACTION_TEMPLATE_ID}}",
            "Sequence": faker.number.int({ max: 9999 }),

            };
            console.log('Search Params:', courseContentSearchData);

            const courseContentSearch = await this.search(courseContentSearchData);
            console.log('Search:', JSON.stringify(courseContentSearch.Data, null, 2));

            const courseContentUpdateData = {
            "Title": faker.lorem.sentence(3),
            "Description": faker.lorem.words(3),
            "ImageUrl": faker.internet.url(),
            "DurationInMins": faker.number.int({ max: 9999 }),
            "ContentType": "Text",
            "ResourceLink": faker.lorem.sentence(3),
            "ActionTemplateId": "{{ACTION_TEMPLATE_ID}}",
            "Sequence": faker.number.int({ max: 9999 }),

            };

            const courseContentUpdate = await this.update(course_contents_Id, courseContentUpdateData);
            console.log('Update:', JSON.stringify(courseContentUpdate.Data, null, 2));

            const courseContentDelete = await this.delete(course_contents_Id);
            console.log('Delete:', JSON.stringify(courseContentDelete.Data, null, 2));

        } catch (error) {
            console.error('Error during course_contents operations:', error);
        }
    }
}

File: sdk/dotnet/lmsservice_db/src/api/course.content/CourseContent.cs

using Newtonsoft.Json;
using Bogus;
using lmsservice_db.src.api.course.content.course.contentvalidation;
using lmsservice_db.src.api.course.content.course.contentmodel;
using FluentValidation;

namespace lmsservice_db.src.api.course.content.course.contentservice;

public class CourseContent
{
    private readonly APIClient _client;

    public CourseContent(APIClient client)
    {
        _client = client;
    }
    public Task<string> Create(CourseContentCreateModel courseContent)
    {
        return _client.Request("/course.content", HttpMethod.Post, courseContent);
    }

    public Task<string> GetById(string courseContentId)
    {
        return _client.Request($"/course.content/{courseContentId}", HttpMethod.Get);
    }

    public Task<string> Search(CourseContentSearchFilters searchFilters)
    {
        var queryParams = new List<string>();

        if (!string.IsNullOrEmpty(searchFilters.Title)) queryParams.Add($"Title={searchFilters.Title}");
if (!string.IsNullOrEmpty(searchFilters.Description)) queryParams.Add($"Description={searchFilters.Description}");
if (!string.IsNullOrEmpty(searchFilters.ImageUrl)) queryParams.Add($"ImageUrl={searchFilters.ImageUrl}");
if (searchFilters.DurationInMins != null) queryParams.Add($"DurationInMins={searchFilters.DurationInMins}");
if (!string.IsNullOrEmpty(searchFilters.ResourceLink)) queryParams.Add($"ResourceLink={searchFilters.ResourceLink}");
if (searchFilters.Sequence != null) queryParams.Add($"Sequence={searchFilters.Sequence}");


        var queryString = string.Join("&", queryParams);
        return _client.Request($"/course.content/search?{queryString}", HttpMethod.Get);
    }

    public Task<string> Update(string courseContentId, CourseContentUpdateModel courseContent)
    {
        return _client.Request($"/course.content/{courseContentId}", HttpMethod.Put, courseContent);
    }

    public Task<string> Delete(string courseContentId)
    {
        return _client.Request($"/course.content/{courseContentId}", HttpMethod.Delete);
    }

    public async Task ExecuteCourseContentOperations()
    {
        var faker = new Faker("en");

        var createValidator = new CourseContentCreateModelValidator();
        var createNewCourseContent = new CourseContentCreateModel
        {
                        Title= faker.Lorem.Sentence(3),
            Description= faker.Lorem.Sentence(),
            ImageUrl= faker.Internet.Url(),
            DurationInMins= faker.Random.Int(1, 10000),
            ContentType= "Assessment",
            ResourceLink= faker.Lorem.Sentence(3),
            ActionTemplateId= "{{ACTION_TEMPLATE_ID}}",
            Sequence= faker.Random.Int(1, 10000),

        };

        var createValidationResult = createValidator.Validate(createNewCourseContent);
        if (!createValidationResult.IsValid)
        {
            foreach (var error in createValidationResult.Errors)
            {
                Console.WriteLine("Validation error:" + error.ErrorMessage);
            }
            return;
        }

        if (!ValidationHelper.TryValidateModel(createNewCourseContent, out var validationResults))
        {
            foreach (var validationResult in validationResults)
            {
                Console.WriteLine(validationResult.ErrorMessage);
            }
            return;
        }

        var createResponse = await Create(createNewCourseContent);
        dynamic createdCourseContent = JsonConvert.DeserializeObject(createResponse);
        Console.WriteLine("Create: " + JsonConvert.SerializeObject(createdCourseContent, Formatting.Indented));

        string courseContentId = createdCourseContent.Data.id.ToString();

        var retrievedResponse = await GetById(courseContentId);
        dynamic retrieved = JsonConvert.DeserializeObject(retrievedResponse);
        Console.WriteLine("GetById: " + JsonConvert.SerializeObject(retrieved, Formatting.Indented));

        var searchValidator = new CourseContentSearchFiltersValidator();
        var searchFilters = new CourseContentSearchFilters
        {
                        Title= createNewCourseContent.Title,
            Description= createNewCourseContent.Description,
            ImageUrl= createNewCourseContent.ImageUrl,
            DurationInMins= createNewCourseContent.DurationInMins,
            ResourceLink= createNewCourseContent.ResourceLink,
            Sequence= createNewCourseContent.Sequence,

        };

        var searchValidationResult = searchValidator.Validate(searchFilters);
        if (!searchValidationResult.IsValid)
        {
            foreach (var error in searchValidationResult.Errors)
            {
                Console.WriteLine("Validation error:" + error.ErrorMessage);
            }
            return;
        }

        var searchResultsResponse = await Search(searchFilters);
        dynamic searchResults = JsonConvert.DeserializeObject(searchResultsResponse);
        Console.WriteLine("Search: " + JsonConvert.SerializeObject(searchResults, Formatting.Indented));

        var updateValidator = new CourseContentUpdateModelValidator();
        var updatedCourseContent = new CourseContentUpdateModel
        {
                        Title= faker.Lorem.Sentence(3),
            Description= faker.Lorem.Sentence(),
            ImageUrl= faker.Internet.Url(),
            DurationInMins= faker.Random.Int(1, 10000),
            ContentType= "Assessment",
            ResourceLink= faker.Lorem.Sentence(3),
            ActionTemplateId= "{{ACTION_TEMPLATE_ID}}",
            Sequence= faker.Random.Int(1, 10000),

        };

        var updateValidationResult = updateValidator.Validate(updatedCourseContent);
        if (!updateValidationResult.IsValid)
        {
            foreach (var error in updateValidationResult.Errors)
            {
                Console.WriteLine("Validation error:" + error.ErrorMessage);
            }
            return;
        }

        var updateResponse = await Update(courseContentId, updatedCourseContent);
        dynamic updated = JsonConvert.DeserializeObject(updateResponse);
        Console.WriteLine("Update: " + JsonConvert.SerializeObject(updated, Formatting.Indented));

        var deleteResponse = await Delete(courseContentId);
        dynamic deleted = JsonConvert.DeserializeObject(deleteResponse);
        Console.WriteLine("Delete: " + JsonConvert.SerializeObject(deleted, Formatting.Indented));
    }
}
